lib/eslint/eslint-helpers.js

```diff
@@ -278,93 +257,107 @@ async function globSearch({
-                     * and so we use `unmatchedPatterns` to do that. There is a bit of
-                     * complexity here because the same file can be matched by more than
-                     * one pattern. So, when we start, we actually need to test every
-                     * pattern against every file. Once we know there are no remaining
-                     * unmatched patterns, then we can switch to just looking for the
-                     * first matching pattern for improved speed.
-                     */
-                    const matchesPattern = unmatchedPatterns.size > 0
-                        ? matchers.reduce((previousValue, matcher) => {
-                            const pathMatches = matcher.match(relativePath);
-
-                            /*
-                             * We updated the unmatched patterns set only if the path
-                             * matches and the file has a config. If the file has no
-                             * config, that means there wasn't a match for the
-                             * pattern so it should not be removed.
-                             *
-                             * Performance note: `getConfig()` aggressively caches
-                             * results so there is no performance penalty for calling
-                             * it multiple times with the same argument.
-                             */
-                            if (pathMatches && configs.getConfig(entry.path)) {
-                                unmatchedPatterns.delete(matcher.pattern);
-                            }
-
-                            return pathMatches || previousValue;
-                        }, false)
-                        : matchers.some(matcher => matcher.match(relativePath));
-
-                    return matchesPattern && configs.getConfig(entry.path) !== void 0;
-                })
+                const absolutePath = normalizeToPosix(path.resolve(basePath, entry.path));
+
+                /*
+                 * If we are in a directory that was passed directly on the command line,
+                 * don't look for a config file. If "subdir" is passed, then we don't want
+                 * to look for a config file in its parent directory. The same applies if "subdir/*"
+                 * "sub*r/*" etc. are passed.
+                 *
+                 * This special case only applies to immediate children of `basePath`, so we
+                 * don't need to worry about any paths that don't include a slash.
+                 *
+                 * Basically, if the raw pattern isn't "." and the entry path doesn't include
+                 * a slash, then we don't need to look for a config file.
+                 */
+                const configs = rawPatterns[matchIndex] !== "." && !entry.path.includes("/")
+                    ? void 0
+                    : await configLoader.loadConfigArrayForDirectory(absolutePath);
```

Comment:
Special-casing only immediate children would cause inconsistent behavior, in my opinion. For example, if the root config file ignores `foo` and `bar/baz` directories, and both of them have their own config files, `eslint foo` would work, while `eslint bar/baz` wouldn't.

Reply:
Sorry, I'm having trouble following what it is you're describing. Can you put together a test case that will fail?

And do you have other suggestions for how you might approach this? It turns out getting the exact behavior in the RFC is pretty difficult and this was the only approach I found that passed all of the tests we had described there.

Reply:
Here's a test case: https://github.com/eslint/eslint/commit/bb439b137c4ee016934895f9f7ee866e7b671916

Reply:
One approach could be to search per each pattern and start loading configs from `globParent()` of the pattern. That's how eslintrc's FileEnumerator works. A downside is performance as the same directories might be traversed multiple times.

Another approach could be to keep searching with multiple patterns at once but check if the directory path is a part of the `globParent()` of at least one of the patterns, and in that case enter the directory even if it's ignored by a config in a parent directory.

Reply:
Here's another test that I think should pass but is failing with the current iteration of this PR: https://github.com/eslint/eslint/commit/7e2193771c8b4f0452c3ec531743b8db0dbd2940

Reply:
Another possible approach: perhaps we could always group patterns by their `globParent()` and start loading configs from that path.